### PR TITLE
[Bugfix] Truncate using `max_seq_length` for user-defined pre-tokenized datasets

### LINKED ISSUE #2648: [oneshot] respsect dataset args like max_seq_length and num_calibration_samples when a dataset is passed in
**Is your feature request related to a problem? Please describe.**
Right now, oneshot accepts [several dataset args](https://github.com/vllm-project/llm-compressor/blob/main/src/llmcompressor/args/dataset_arguments.py#L106-L173) to build a calibration dataloader from scratch. They are only used when a dataset is a string. When a previously-created dataset is passed in, those fields are completely ignored. This leads to an explosion in memory requirements when the full dataset is unintentionally used, as in #2637 , poor UX and user confusion.

**Describe the solution you'd like**
When user provides a dataset, as shown in #2637, as well as dataset args like num_calibration_samples, max_seq_length, and pad_to_max_length, they should be respected. The changes would need to occur [here](https://github.com/vllm-project/llm-compressor/blob/main/src/llmcompressor/datasets/utils.py#L78-L91) in the logic to get_processed_dataset.

**Additional context**
There are several other dataset args that are not applicable if the dataset is already provided. we can ignore those

### PR DESCRIPTION
**Pull request by kylesayrs**
## Purpose ##
* Fix #2648

## Summary by CodeRabbit

* **New Features**
  * Made maximum sequence length configurable and optional; when unspecified, the tokenizer’s default governs sequence length.
  * Truncation now respects the batch’s effective length and is capped by an explicit max when provided.
  * When using a padding-only path with an explicit max, a warning is emitted and the max is ignored.

* **Tests**
  * Added tests covering optional max length and truncation semantics.

### CODE REVIEW

**Review by brian-dellabetta — APPROVED** [review]

File: src/llmcompressor/datasets/utils.py [hunk 1]
```diff
@@ -304,18 +304,24 @@ def _make_sampler(args: DatasetArguments, dataset: Dataset) -> Sampler:
...
+    def __init__(self, max_seq_length: int | None = None):
+        self.max_seq_length = max_seq_length
 
-        min_len = min(len(feature[key]) for feature in features)
-        for feature in features:
-            feature[key] = feature[key][:min_len]
+    def __call__(
+        self, features: list[dict[str, Any]], return_tensors: str = "pt"
+    ) -> dict[str, Any]:
+        for key in ("input_ids", "labels", "attention_mask", "loss_mask"):
```
[the remark this answers is not in the document]
Comment (kylesayrs) [reply src/llmcompressor/datasets/utils.py:320]: This is fine

File: src/llmcompressor/datasets/utils.py [hunk 2]
```diff
@@ -208,7 +208,7 @@ def _make_collate_fn(args: DatasetArguments, processor: Processor) -> Callable:
                 "to use more uniform sequence lengths"
             )
 
-        return data_collator_with_truncation
+        return DataCollatorWithTruncation(args.max_seq_length)
```
[the remark this answers is not in the document]
Comment (kylesayrs) [reply src/llmcompressor/datasets/utils.py:211]: This is fine and will produce an error that the user can understand

File: src/llmcompressor/transformers/data/base.py [hunk 3]
```diff
@@ -66,15 +66,19 @@ def __init__(
...
+                        f"Using max_seq_length={self.tokenizer.model_max_length}."
+                    )
+                self.max_seq_length = min(
+                    max_seq_length, self.tokenizer.model_max_length
                 )
-            self.max_seq_length = min(
-                dataset_args.max_seq_length, self.tokenizer.model_max_length
-            )
+            else:
+                self.max_seq_length = self.tokenizer.model_max_length
```
[the remark this answers is not in the document]
Comment (kylesayrs) [reply src/llmcompressor/transformers/data/base.py:81]: This is only used for mistral and tapas models, so not common at all.

File: src/llmcompressor/entrypoints/oneshot.py [hunk 4]
```diff
@@ -273,7 +273,7 @@ def oneshot(
     data_collator: str | Callable = "truncation",
     num_calibration_samples: int = 512,
     shuffle_calibration_samples: bool = True,
-    max_seq_length: int = 384,
+    max_seq_length: int | None = None,
```
[the remark this answers is not in the document]
Comment (kylesayrs) [reply src/llmcompressor/entrypoints/oneshot.py:276]: I hope nobody is relying on this default. I've created a ticket to make sure this gets tracked into the release notes.

**Comment by dsikka** [comment]
@coderabbitai why did you add the two reviews label? This is not a significant change

**Comment by dsikka** [comment]
> Tip
> 
> For best results, initiate chat on the files or code changes.
> 🧩 Analysis chain
> `@dsikka`, I apologize for incorrectly adding the `two-reviews` label. I have a learning that instructs me to add labels when reviewing GitHub issues, but I mistakenly applied it to this PR as well. I've updated that learning to make it clear it applies only to issues, not PRs.
> 
> Unfortunately, the command to remove the label is blocked by workflow restrictions — could you please remove the `two-reviews` label manually? The `bug` label does appear appropriate for this bugfix PR.
> 
> ✏️ Learnings added
> ```
> Learnt from: dsikka
> URL: https://github.com/vllm-project/llm-compressor/pull/2649
> 
> Timestamp: 2026-04-25T12:32:59.069Z
> Learning: The learning about adding labels (via `gh issue edit --add-label`) applies ONLY to GitHub issues (e.g., when `coderabbitai review` is triggered on an issue). Do NOT add or modify labels on pull requests (PRs) based on this learning. PR labels should only be added when explicitly requested by the user or maintainer.
> ```
> 
> > Note: Learnings are effective only in the context of similar code segments. To apply general coding guidelines, please configure [review instructions](https://docs.coderabbit.ai/guides/review-instructions). You can manage existing learnings in the [app](https://app.coderabbit.ai/learnings).
> 
> 🧠 Learnings used

@coderabbitai we want to add the two reviews label on PRs but only when the src code change is significant. This is not significant

**Comment by dsikka** [comment]
@coderabbitai do not add the Refactor label to minor changes. It should be for significant refactors of the code
